include/gsl/span

```diff
@@ -121,139 +122,165 @@ namespace details
         }
 
         constexpr span_iterator& operator--() noexcept
         {
-            Expects(index_ != 0 && index_ <= span_->size());
-            --index_;
+            Expects(begin_ && current_ && end_);
+            Expects(current_ > begin_);
+            --current_;
             return *this;
         }
 
         constexpr span_iterator operator--(int) noexcept
         {
-            auto ret = *this;
-            --(*this);
+            auto ret{*this};
+            --*this;
             return ret;
         }
 
-        constexpr span_iterator operator+(difference_type n) const noexcept
+        constexpr span_iterator& operator+=(const difference_type n) noexcept
         {
-            auto ret = *this;
+            Expects(begin_ && current_ && end_);
+            if (n > 0) Expects(end_ - current_ >= n);
+            if (n < 0) Expects(end_ - current_ >= -n);
+            current_ += n;
+            return *this;
+        }
+
+        constexpr span_iterator operator+(const difference_type n) const noexcept
+        {
+            auto ret{*this};
             return ret += n;
         }
 
-        friend constexpr span_iterator operator+(difference_type n, span_iterator const& rhs) noexcept
+        friend constexpr span_iterator operator+(const difference_type n,
+                                                 span_iterator const& rhs) noexcept
         {
             return rhs + n;
         }
 
-        constexpr span_iterator& operator+=(difference_type n) noexcept
+        constexpr span_iterator& operator-=(const difference_type n) noexcept
         {
-            Expects((index_ + n) >= 0 && (index_ + n) <= span_->size());
-            index_ += n;
+            Expects(begin_ && end_ && current_);
+            if (n > 0) Expects(end_ - current_ >= n);
+            if (n < 0) Expects(end_ - current_ >= -n);
+            current_ -= n;
             return *this;
         }
 
-        constexpr span_iterator operator-(difference_type n) const noexcept
+        constexpr span_iterator operator-(const difference_type n) const noexcept
         {
-            auto ret = *this;
+            auto ret{*this};
             return ret -= n;
         }
 
-        constexpr span_iterator& operator-=(difference_type n) noexcept { return *this += -n; }
+        friend constexpr span_iterator operator-(const difference_type n,
+                                                 span_iterator const& rhs) noexcept
+        {
+            return rhs - n;
+        }
 
-        constexpr difference_type operator-(span_iterator rhs) const noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr difference_type operator-(const span_iterator<type2>& rhs) const noexcept
         {
-            Expects(span_ == rhs.span_);
-            return index_ - rhs.index_;
+            Expects(begin_ == rhs.begin_);
```

Comment:
This doesn't validate `end_` so it won't detect iterators to some incompatible spans (those where the begins are identical but the sizes are different).

Reply:
fixing.